src/ExecutionContext/index.js

```diff
@@ -1,175 +1,69 @@
-const asyncHooks = require('async_hooks');
-const { isProduction, monitorMap, ExecutionContextResource } = require('../lib');
-const { create: createHooks, onChildProcessDestroy } = require('../hooks');
-const {
-    DEFAULT_CONFIG,
-    ROOT_DOMAIN,
-    ExecutionContextErrors
-} = require('./constants');
-
-/**
- * The execution context maps which acts as the execution context in memory storage.
- * @see ExecutionContext.monitor
- * @type ExecutionContextMap
- */
-const executionContextMap = new Map();
-
-/**
- * Creates a root execution context node.
- * @param {Number} asyncId - The current async id.
- * @param {Object} initialContext - The initial context ro provide this execution chain.
- * @param {Object} config - The configuration the root context created with.
- * @param {Object} domain - The domain to create the execution context under.
- * @return {ExecutionContextNode}
- */
-const createRootContext = ({ asyncId, initialContext, config, domain = ROOT_DOMAIN }) => ({
-    asyncId,
-    domain,
-    context: { ...initialContext, executionId: asyncId },
-    children: [],
-    ...config,
-    ...(config.monitor && { created: Date.now() })
-});
-
-/**
- * Handles execution context error, throws when none production.
- * @param {String} code - The error code to log.
- */
-const handleError = (code) => {
-    if (!isProduction()) {
-        throw code;
-    }
-
-    console.error(code); // eslint-disable-line no-console
-};
+const { supportAsyncLocalStorage } = require('../lib');
+const { AsyncHooksContext, AsyncLocalStorageContext } = require('../managers');
 
 /**
  * @type {ExecutionContextAPI}
  */
 class ExecutionContext {
     constructor() {
-        this.config = { ...DEFAULT_CONFIG };
-
-        // Sets node async hooks setup
-        asyncHooks.createHook(
-            createHooks(executionContextMap)
-        ).enable();
-    }
+        const ContextManager = supportAsyncLocalStorage() ? AsyncLocalStorageContext : AsyncHooksContext;
```

Comment:
this is cool and i liked the abstraction, but did you consider creating an entirely new release version for node 12.7 and above? pros cons?